spikeinterface/sortingcomponents/dimensionality_reduction.py

```diff
@@ -0,0 +1,141 @@
+            with open(model_path, "rb") as f:
+                self.pca_model = pickle.load(f)
+        
+
+    def get_dtype(self):
+        return self._dtype
+
+    def fit(self, recording: BaseRecording, n_components: int, detect_peaks_params: dict, 
+            peak_selection_params: dict, whiten: bool = True, **job_kwargs) -> IncrementalPCA:
+        """
+        Train a pca model using the data in the recording object and the parameters provided.
+        Note that this model returns the pca model from scikit-learn but the model is also saved in the path provided
+        when instantiating the class.
+        
+        Parameters
+        ----------
+        recording : BaseRecording
+            The recording object.
+        n_components : int
+            The number of components to use for the PCA model.
+        detect_peaks_params : dict
+            The parameters for peak detection.
+        peak_selection_params : dict
+            The parameters for peak selection.
+        whiten : bool, optional
+            Whether to whiten the data, by default True.
+
+        {}
+        
+        Returns
+        -------
+        IncrementalPCA: An incremental PCA estimator from scikit-learn.
+            The pca model
+        """
+        
+        # Detect peaks and sub-sample them
+        peaks = detect_peaks(recording, **detect_peaks_params, **job_kwargs)
+        peaks = select_peaks(peaks, **peak_selection_params) # How to select n_peaks
+
+        # Create a waveform extractor
+        ms_before = self._kwargs["ms_before"]
+        ms_after = self._kwargs["ms_after"]
+        # Creates a numpy sorting object where the spike times are the peak times and the unit ids are the peak channel
+        sorting = NumpySorting.from_peaks(peaks, sampling_frequency=recording.sampling_frequency) 
+        we = extract_waveforms(recording, sorting, ms_before=ms_before, ms_after=ms_after, folder=None, 
+                               mode="memory", max_spikes_per_unit=None, **job_kwargs)
+
+        # compute PCA by_channel_global (with sparsity)
+        sparsity = ChannelSparsity.from_radius(we, radius_um=self.local_radius_um)
+        pc = compute_principal_components(we, n_components=n_components,  mode='by_channel_global',
+                                        sparsity=sparsity, whiten=whiten)
+        self.pca_model  = pc.get_pca_model()
+
+        # model_folder should be an input
+        pca_model_path = self._kwargs["model_path"]
+
+        with open(pca_model_path, "wb") as f:
+            pickle.dump(self.pca_model, f)
+            
+        return self.pca_model
+    
+    def compute_buffer(self, traces, peaks, waveforms):        
+
+        if self.pca_model == None:
+            exception_string = (
+                f"Pca model not found, "
+                f"Train the pca model before running compute_buffer using {self.__name__}.fit(kwargs)"
+            )
+            raise AttributeError(exception_string)
```

Comment:
This check should be done before.
Here this is call at every chunk.

Reply:
I don't see where else this could be. You can not do it in the __init__ because you might call `fit` after initialization.

Reply:
I agree with Ramon. this only adds the `if self.pca_model == None:` in terms of computation and we need to protect against running the compute buffer without an instantiated pca model (which is possible at init if you instantiate the component for fitting!)